modin/backends/pandas/query_compiler.py

```diff
@@ -1009,6 +1009,73 @@ def rolling_aggregate(self, rolling_args, func, *args, **kwargs):
         )
         return self.__constructor__(new_modin_frame)
 
+    def unstack(self, level, fill_value):
+        if not isinstance(self.index, pandas.MultiIndex) or (
+            isinstance(self.index, pandas.MultiIndex)
+            and is_list_like(level)
+            and len(level) == self.index.nlevels
+        ):
+            axis = 1
+            new_columns = ["__reduced__"]
+            need_reindex = True
+        else:
+            axis = 0
+            new_columns = None
+            need_reindex = False
+
+        def map_func(df):
+            return pandas.DataFrame(df.unstack(level=level, fill_value=fill_value))
+
+        is_all_multi_list = False
+        if (
+            isinstance(self.index, pandas.MultiIndex)
+            and isinstance(self.columns, pandas.MultiIndex)
+            and is_list_like(level)
+            and len(level) == self.index.nlevels
+        ):
+            is_all_multi_list = True
+            real_cols_bkp = self.columns
+            obj = self.copy()
+            obj.columns = np.arange(len(obj.columns))
+        else:
+            obj = self
+
+        new_modin_frame = obj._modin_frame._apply_full_axis(
+            axis, map_func, new_columns=new_columns
+        )
+        result = self.__constructor__(new_modin_frame)
+
+        if is_all_multi_list:
+            result = result.sort_index()
+
+            real_cols_extended = [
+                col for col in real_cols_bkp for _ in np.arange(len(obj.index))
+            ]
+            reverted_index = [
+                col + tup[1:] for col, tup in zip(real_cols_extended, result.index)
+            ]
+            result.index = pandas.MultiIndex.from_tuples(reverted_index)
+            return result
+
+        if need_reindex:
+            if isinstance(self.index, pandas.MultiIndex):
+                new_tuples = [
+                    [val] + list(tup) for val in self.columns for tup in self.index
+                ]
```

Comment:
Am I correct that this loop has O(N*M) complexity in case of (N, M) source DataFrame? That loop can cost us mush in case of big dimensional DataFrames, can we somehow reduce the complexity of that loop?

Reply:
It was fixed, at now all loops has O(nlevels) complexity.